lib/jxl/dec_cache.cc

```diff
@@ -195,23 +196,57 @@ Status PassesDecoderState::PreparePipeline(ImageBundle* decoded,
       }
     }
 
+    ColorEncoding linear_color_encoding;
+    if (frame_header.color_transform == ColorTransform::kXYB) {
+      linear_color_encoding = output_encoding_info.linear_color_encoding;
+    } else {
+      linear_color_encoding = ColorEncoding::LinearSRGB(
+          output_encoding_info.orig_color_encoding.IsGray());
+    }
+
     auto tone_mapping_stage = GetToneMappingStage(output_encoding_info);
     if (tone_mapping_stage) {
       if (!linear) {
         auto to_linear_stage = GetToLinearStage(output_encoding_info);
         if (!to_linear_stage) {
-          return JXL_FAILURE(
-              "attempting to perform tone mapping on colorspace not "
-              "convertible to linear");
+          if (!output_encoding_info.color_management_system) {
+            return JXL_FAILURE("Cannot tonemap this colorspace without a CMS");
+          }
+          auto cms_stage = GetCmsStage(output_encoding_info);
+          if (cms_stage) {
+            builder.AddStage(std::move(cms_stage));
+          }
+        } else {
+          builder.AddStage(std::move(to_linear_stage));
         }
-        builder.AddStage(std::move(to_linear_stage));
         linear = true;
       }
       builder.AddStage(std::move(tone_mapping_stage));
     }
 
     if (linear) {
-      builder.AddStage(GetFromLinearStage(output_encoding_info));
+      const size_t channels_src =
+          (output_encoding_info.linear_color_encoding.IsCMYK()
```

Comment:
... For that matter, what's the linear_color_encoding supposed to be in the CMYK case?

Reply:
no idea. @sboukortt ?

Reply:
The `linear_color_encoding` is whatever is output from the from_xyb stage, also in the CMYK stage.
I added a comment to the new field `linear_color_encoding` to explain what it is.